doc/md/canister-maintenance/compatibility.md

```diff
@@ -158,6 +168,73 @@ For this purpose, a user-instructed migration can be done in three steps:
 
 Alternatively, the type of `state` can be changed to `Any`, also implying that this variable is no longer used.
 
+### Explicit migration using a migration function
+
+The previous approach of using several upgrades to migrate data is both tedious and
+obscure, mingling production with migration code.
+
+To ease data migration, Motoko now supports explicit migration using a separate data migration function.
+The code for the migration function is self-contained and can be placed in its own file.
+
+The migration function takes a record of stable fields as input and produces a record of stable fields as output.
+
+The input fields extend or override the types of any stable fields in the actor's
+stable signature.
+The output fields must be declared in the actors stable signature, and have types that can be consumed by the corresponding declaration in the stable signature.
+
+* All values for the input fields must
+be present and of compatible type in the old actor, otherwise the
+upgrade traps and rolls back.
+* The fields output by the migration
+function determine the values of the corresponding stable variables in the
+new actor.
+* All other stable variables of the actor, i.e. those neither consumed nor
+produced by the migration function are initialized in the usual way,
+either by transfer from the upgraded actor, if declared in that actor, or, if newly declared,
+by running the initialization expression in the field's
+declaration.
+* The migration function is only executed on an upgrade and ignored on a fresh installation of the actor in an empty canister.
+
+The migration function, when required, is declared within square brackets following the `actor` keyword in an actor or actor class
+declaration, for example:
```

Comment:
Any chance that (later) we adopt the `(with migration = ...)` syntax?

Reply:
Done